torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,240 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, Optional, Tuple
+
+import torch
+from torch import nn, Tensor
+from torch.nn import functional as F
+from torchmultimodal.utils.common import shift_dim
+
+
+class MultiHeadAttention(nn.Module):
+    """Compute multihead attention with flexible attention function.
+
+    Multihead attention linearly projects queries, keys, and values into an embedding
+    space, which is divided into multiple 'heads'. Attention is computed for each head
+    individually instead of across the entire embedding space once. This enables more
+    varied representations and allows the model to jointly attend to information from
+    different representation subspaces at different positions, as described in
+    Attention Is All You Need (Vaswani et al. 2017).
+
+    Args:
+        shape (Tuple): shape of input data (d1 x ... x dn)
+        dim_q (int): dimensionality of input into query weights
+        dim_kv (int): dimensionality of input into key and value weights
+        n_head (int): number of attention heads
+        n_layer (int): ?
```

Comment:
`n_layer` is the number of attention blocks specified by the GPT transformer. I'm not sure why it's used by the feed forward layer for a single attention block. Could you post a question to Songyang?